src/simulators/statevector/qv_avx2.cpp

```diff
@@ -1118,21 +1171,24 @@ Avx apply_diagonal_matrix_avx<double>(double* qv_data_,
                                       const double* vec_,
                                       const size_t omp_threads) {
 
-  std::complex<double>* qv_data = reinterpret_cast<std::complex<double>*>(qv_data_);
-  const std::complex<double>* input_vec = reinterpret_cast<const std::complex<double>*>(vec_);
+  auto qv_data = _to_complex(qv_data_);
+  const auto input_vec = _to_complex(vec_);
 
-  std::complex<double>** inputs;
-  inputs = reinterpret_cast<std::complex<double>**>(malloc(sizeof(std::complex<double*>) * omp_threads));
+  std::complex<double>** tmps;
+  tmps = reinterpret_cast<std::complex<double>**>(malloc(sizeof(std::complex<double>*) * omp_threads));
 
-#pragma omp parallel for if (omp_threads > 1) num_threads(omp_threads)
-  for (int i = 0; i < omp_threads; ++i) {
+#pragma omp parallel if (omp_threads > 1) num_threads(omp_threads)
+  {
+#pragma omp for
+    for (int i = 0; i < omp_threads; ++i) {
```

Comment:
Not sure but, aren't these statements redundant? Aren't we already parallelizing in `num_threads` with the above `#pragma omp parallel if (omp_threads > 1) num_threads(omp_threads)` statement?

Reply:
This is the same with [`indexes.hpp`](https://github.com/Qiskit/qiskit-aer/blob/a8b30d3e07129cbedeb4ded9cb03e18132fb542f/src/simulators/statevector/indexes.hpp#L180-L187). `omp parallel` is applied to `for`-loop with `#pragma omp for` annotation within its scope, in my understanding.

Reply:
But in `indexes.hpp` we use the index `k` coming from the `for` statement and running from `start` to `stop`. Here we don't use the index from the `for` in any way and we are running the loop from 0 to `num_threads`. Looks like with the first `pragma` we are already creating the threads that we need.

Reply:
Oh, that's true. In this case, we do not need `omp` for `for`. Thanks for your comments!